tempesta_fw/apm.c

```diff
@@ -1058,6 +1213,197 @@ tfw_apm_del_srv(TfwServer *srv)
+		 */
+		if (hmstats[i].hmcfg->code == status ||
+		    hmstats[i].hmcfg->code == status / 100)
+		{
+			history = hmstats[i].history;
+			cfg = hmstats[i].hmcfg;
+			hmstats = &hmstats[i];
+			break;
+		}
+	}
+
+	if (!history)
+		return false;
+
+	BUG_ON(!cfg);
+	spin_lock(&hmstats->lock);
+
+	ts = jiffies * HM_FREQ / (cfg->tframe * HZ);
+	i = ts % HM_FREQ;
+	if (history[i].ts != ts) {
+		history[i].ts = ts;
+		history[i].resp = 0;
+	}
+	++history[i].resp;
+	for (i = 0; i < HM_FREQ; ++i)
+		if (history[i].ts + HM_FREQ > ts)
+			sum += history[i].resp;
+
+	WRITE_ONCE(hmstats->rsum, sum);
+	spin_unlock(&hmstats->lock);
+
+	if (sum > cfg->limit)
+		return true;
+
+	return false;
+}
+
+void
+tfw_apm_hm_enable_srv(TfwServer *srv, void *hmref)
+{
+	TfwApmHMCtl *hmctl;
+	unsigned long now;
+	TfwApmHM *hm = hmref;
+
+	BUG_ON(!srv->apmref);
+	BUG_ON(!hm);
+	BUG_ON(test_bit(TFW_SRV_B_HMONITOR, &srv->hm_flags));
+
+	/* Set new health monitor for server. */
+	hmctl = &((TfwApmData *)srv->apmref)->hmctl;
+	WRITE_ONCE(hmctl->hm, hm);
+
+	/* Init server's health control fields. */
+	atomic64_set(&hmctl->rcount, 0);
+	clear_bit(TFW_SRV_B_SUSPEND, &srv->hm_flags);
+
+	/* Start server's health monitoring timer. */
+	atomic_set(&hmctl->rearm, 1);
+	smp_mb__after_atomic();
+	setup_timer(&hmctl->timer, tfw_apm_hm_timer_cb, (unsigned long)srv);
+	now = jiffies;
+	mod_timer(&hmctl->timer, now + hm->tmt * HZ);
+	WRITE_ONCE(hmctl->jtmstamp, now);
+
+	/* Activate server's health monitor. */
+	set_bit(TFW_SRV_B_HMONITOR, &srv->hm_flags);
+}
+
+void
+tfw_apm_hm_disable_srv(TfwServer *srv)
+{
+	clear_bit(TFW_SRV_B_HMONITOR, &srv->hm_flags);
+	tfw_apm_hm_stop_timer((TfwApmData *)srv->apmref);
+}
+
+bool
+tfw_apm_hm_srv_eq(const char *name, TfwServer *srv)
+{
+	TfwApmHM *hm;
+
+	BUG_ON(!srv->apmref);
+	hm = ((TfwApmData *)srv->apmref)->hmctl.hm;
+	BUG_ON(!hm);
+	if(!strcasecmp(name, hm->name))
+		return true;
+
+	return false;
+}
+
+/*
+ * Calculation of general health monitoring statistics (for procfs).
+ * Function allocates new TfwHMStats object which must be freed by
+ * the caller.
+ */
+TfwHMStats *
+tfw_apm_hm_stats(void *apmref)
+{
+	int i;
+	long rtime;
+	size_t size;
+	TfwHMStats *stats;
+	TfwHMCodeStats *rsums;
+	TfwApmHMCtl *hmctl = &((TfwApmData *)apmref)->hmctl;
+	TfwApmHM *hm = READ_ONCE(hmctl->hm);
+
+	BUG_ON(!hmctl->hmstats || !hm);
+	size = sizeof(TfwHMStats) + sizeof(TfwHMCodeStats) * tfw_hm_codes_cnt;
+	if (!(stats = kmalloc(size, GFP_KERNEL)))
+		return NULL;
+
+	rsums = (TfwHMCodeStats *)(stats + 1);
+	for (i = 0; i < tfw_hm_codes_cnt; ++i) {
+		BUG_ON(!hmctl->hmstats[i].hmcfg);
+		rsums[i].code = hmctl->hmstats[i].hmcfg->code;
+		rsums[i].sum = READ_ONCE(hmctl->hmstats[i].rsum);
+	}
+	stats->rsums = rsums;
+	stats->ccnt = tfw_hm_codes_cnt;
+
+	rtime = (long)hm->tmt - (jiffies - READ_ONCE(hmctl->jtmstamp))
+					/ HZ;
+	stats->rtime = rtime < 0 ? 0 : rtime;
+
+	return stats;
+}
+
+bool
+tfw_apm_get_hm(const char *name, void **res_hm)
```

Comment:
Why not to return `hm`? Why do we need the pointer and a boolean return value?

Reply:
Corrected.